presto-main/src/main/java/com/facebook/presto/operator/FileFragmentResultCacheManager.java

```diff
@@ -0,0 +1,329 @@
+        if (fragmentCacheStats.getInFlightBytes() + resultSize > maxInFlightBytes ||
+                cache.getIfPresent(key) != null ||
+                resultSize > maxSinglePagesBytes ||
+                // Here we use the logical size resultSize as an estimate for admission control.
+                fragmentCacheStats.getCacheSizeInBytes() + resultSize > maxCacheBytes) {
+            return immediateFuture(null);
+        }
+
+        fragmentCacheStats.addInFlightBytes(resultSize);
+        Path path = baseDirectory.resolve(randomUUID().toString().replaceAll("-", "_"));
+        return flushExecutor.submit(() -> cachePages(key, path, result, resultSize));
+    }
+
+    private static long getPagesSize(List<Page> pages)
+    {
+        return pages.stream()
+                .mapToLong(Page::getSizeInBytes)
+                .sum();
+    }
+
+    private void cachePages(CacheKey key, Path path, List<Page> pages, long resultSize)
+    {
+        try {
+            Files.createFile(path);
+            try (SliceOutput output = new OutputStreamSliceOutput(newOutputStream(path, APPEND))) {
+                writePages(pagesSerdeFactory.createPagesSerde(), output, pages.iterator());
+                long resultPhysicalBytes = output.size();
+                cache.put(key, new CacheEntry(path, resultPhysicalBytes));
+                fragmentCacheStats.incrementCacheEntries();
+                fragmentCacheStats.addCacheSizeInBytes(resultPhysicalBytes);
+            }
+            catch (UncheckedIOException | IOException e) {
+                log.warn(e, "%s encountered an error while writing to path %s", Thread.currentThread().getName(), path);
+                tryDeleteFile(path);
+            }
+        }
+        catch (UncheckedIOException | IOException e) {
+            log.warn(e, "%s encountered an error while writing to path %s", Thread.currentThread().getName(), path);
+            tryDeleteFile(path);
+        }
+        finally {
+            fragmentCacheStats.addInFlightBytes(-resultSize);
+        }
+    }
+
+    private static void tryDeleteFile(Path path)
+    {
+        try {
+            File file = new File(path.toUri());
+            if (file.exists()) {
+                Files.delete(file.toPath());
+            }
+        }
+        catch (IOException e) {
+            // ignore
+        }
+    }
+
+    @Override
+    public Optional<Iterator<Page>> get(String serializedPlan, Split split)
+    {
+        CacheKey key = new CacheKey(serializedPlan, split.getSplitIdentifier());
+        CacheEntry cacheEntry = cache.getIfPresent(key);
+        if (cacheEntry == null) {
+            fragmentCacheStats.incrementCacheMiss();
+            return Optional.empty();
+        }
+
+        try {
+            InputStream inputStream = newInputStream(cacheEntry.getPath());
+            Iterator<Page> result = readPages(pagesSerdeFactory.createPagesSerde(), new InputStreamSliceInput(inputStream));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code contains a resource that might not have closed properly. A resource leak can slow down or crash your system.

**Fix**
Consider closing the resource returned by the *InputStreamSliceInput* constructor in a try-finally block. The resource is referenced at line 210. The resource is closed at line 212. There are other execution paths that do not contain closure statements, for example, when FragmentCacheStats.incrementCacheHit throws an exception.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).